### PR TITLE
3.5 streamed replicated content

### PR DESCRIPTION
**Pull request by RagnarW**
Improves performance for large transactions by chunking data. With default netty configs this gives major improvements for tx > 16mb

Marshaling is unchanged meaning what we write and read to raft log is the same.

Raft messages are split up in Messages, Raft log entry terms and Replicated Content. Replicated content is always a chunked input but really only used by transactions and dummy requests.

### CODE REVIEW

**Review by martinfurmanski — COMMENTED** [review]
I think there is too much duplication going on and we should be able to align the code much more. It will make it clearer also what the differences are going from V1 to V2. This is just an intermediary comment, and I will continue looking at the code as is.

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/EnterpriseCoreEditionModule.java [hunk 1]
```diff
@@ -268,9 +268,10 @@ public EnterpriseCoreEditionModule( final PlatformModule platformModule,
         ModifierProtocolRepository modifierProtocolRepository =
                 new ModifierProtocolRepository( Protocol.ModifierProtocols.values(), supportedModifierProtocols );
 
-        ProtocolInstallerRepository<ProtocolInstaller.Orientation.Client> protocolInstallerRepository =
-                new ProtocolInstallerRepository<>(
-                        singletonList( new RaftProtocolClientInstaller.Factory( clientPipelineBuilderFactory, logProvider ) ),
+        ProtocolInstallerRepository<ProtocolInstaller.Orientation.Client> protocolInstallerRepository = new ProtocolInstallerRepository<>(
+                asList( new RaftProtocolClientInstaller.Factory( clientPipelineBuilderFactory, logProvider ),
+                        new org.neo4j.causalclustering.core.consensus.protocol.v1.RaftProtocolClientInstaller.Factory( clientPipelineBuilderFactory,
```
Comment (martinfurmanski) [inline enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/EnterpriseCoreEditionModule.java]: You can rename the class to have V1 at the end, and avoid this long import name. The new class can have V2. The comment applies to this but also other classes.

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/CoreReplicatedContentSerializer.java [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+import org.neo4j.causalclustering.core.state.machines.tx.ReplicatedTransactionSerializer;
+import org.neo4j.causalclustering.core.state.storage.SafeChannelMarshal;
+import org.neo4j.causalclustering.messaging.EndOfStreamException;
+import org.neo4j.storageengine.api.ReadableChannel;
+import org.neo4j.storageengine.api.WritableChannel;
+
+import static java.util.Collections.singleton;
+import static org.neo4j.causalclustering.messaging.marshalling.Serializer.simple;
+
+public class CoreReplicatedContentSerializer extends SafeChannelMarshal<ReplicatedContent>
```
Comment (martinfurmanski) [inline enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/CoreReplicatedContentSerializer.java]: Why rename from Marshal->Serializer? Makes it harder to code review and also inconsistent with the rest of things.

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/Serializer.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+import io.netty.buffer.ByteBuf;
+
+import java.io.IOException;
+
+import org.neo4j.causalclustering.messaging.NetworkFlushableChannelNetty4;
+import org.neo4j.function.ThrowingConsumer;
+import org.neo4j.storageengine.api.WritableChannel;
+
+public interface Serializer extends Marshal
```
Comment (martinfurmanski) [inline enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/Serializer.java]: These seems like a weird hierarchy and/or naming. I think the way it's been used previously these are basically synonymous. Needs better naming and a Javadoc to begin with.

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/v2/decoding/RaftMessageComposer.java [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+import java.util.Queue;
+import java.util.concurrent.LinkedBlockingQueue;
+
+import org.neo4j.causalclustering.core.consensus.RaftMessages;
+import org.neo4j.causalclustering.core.replication.ReplicatedContent;
+
+public class RaftMessageComposer extends MessageToMessageDecoder<Object>
+{
+    private final Queue<ReplicatedContent> replicatedContents = new LinkedBlockingQueue<>();
+    private final Queue<Long> raftLogEntries = new LinkedBlockingQueue<>();
```
Comment (martinfurmanski) [inline enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/v2/decoding/RaftMessageComposer.java]: Should this be named terms?

File: enterprise/causal-clustering/src/test/java/org/neo4j/causalclustering/core/consensus/log/RaftContentByteBufferMarshalTest.java [hunk 5]
```diff
@@ -1,150 +0,0 @@
...
-import org.neo4j.kernel.impl.transaction.TransactionRepresentation;
-import org.neo4j.kernel.impl.transaction.log.PhysicalTransactionRepresentation;
-import org.neo4j.storageengine.api.StorageCommand;
-
-import static org.hamcrest.CoreMatchers.equalTo;
-import static org.hamcrest.MatcherAssert.assertThat;
-import static org.junit.Assert.assertEquals;
-import static org.neo4j.helpers.collection.Iterators.asSet;
-
-public class RaftContentByteBufferMarshalTest
```
Comment (martinfurmanski) [inline enterprise/causal-clustering/src/test/java/org/neo4j/causalclustering/core/consensus/log/RaftContentByteBufferMarshalTest.java:55]: Did this move somewhere else? I wonder if some of the behaviour should still be captured by a test.

File: enterprise/causal-clustering/src/test/java/org/neo4j/causalclustering/catchup/tx/ChunkedReplicatedTransactionInputStreamTest.java [hunk 6]
```diff
@@ -41,7 +41,7 @@
 import static org.neo4j.causalclustering.catchup.CatchupResult.SUCCESS_END_OF_STREAM;
 import static org.neo4j.kernel.impl.transaction.log.TransactionIdStore.BASE_TX_ID;
 
-public class ChunkedTransactionStreamTest
+public class ChunkedReplicatedTransactionInputStreamTest
```
Comment (martinfurmanski) [inline enterprise/causal-clustering/src/test/java/org/neo4j/causalclustering/catchup/tx/ChunkedReplicatedTransactionInputStreamTest.java]: Seems like an unnecessary and overly verbose rename.

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/ContentBuilder.java [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+ * licensing@neo4j.com
+ *
+ * More information is also available at:
+ * https://neo4j.com/licensing/
+ */
+package org.neo4j.causalclustering.messaging.marshalling;
+
+import java.util.function.Function;
+
+public class ContentBuilder<CONTENT>
```
Comment (martinfurmanski) [inline enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/ContentBuilder.java]: Add JavaDoc please explaining what this does.

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/Marshal.java [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+ * More information is also available at:
+ * https://neo4j.com/licensing/
+ */
+package org.neo4j.causalclustering.messaging.marshalling;
+
+import java.io.IOException;
+
+import org.neo4j.storageengine.api.WritableChannel;
+
+public interface Marshal
```
Comment (martinfurmanski) [inline enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/Marshal.java:29]: Why is this necessary alongside the other Marshal interfaces, e.g. ChannelMarshal? Maybe name it OutboundMarshal since it seems to be unidirectional? Perhaps the regular Marshal-hierarchy should build on this?

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/Serializer.java [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+{
+    /** May override buffer allocation size.
```
Comment (martinfurmanski) [inline enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/Serializer.java]: Please explain what it does and not just a single side-effect.

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/Serializer.java [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+        return new Serializer()
+        {
+            private boolean consumed;
+
+            @Override
+            public boolean encode( ByteBuf byteBuf ) throws IOException
+            {
+                if ( consumed )
+                {
+                    return false;
```
Comment (martinfurmanski) [inline enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/Serializer.java]: This seems like a bit of surprising logic to have in a "serializer". Shouldn't it just serialize?

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/v2/decoding/DecodingDispatcher.java [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+        register( ContentType.RaftLogEntries, new RaftLogEntryTermsDecoder( protocol ) );
+        decoder = new ReplicatedContentChunkDecoder();
+        register( ContentType.ReplicatedContent, decoder );
+        register( ContentType.Message, new RaftMessageDecoder( protocol ) );
+    }
+
+    @Override
+    public void close()
+    {
+        decoder.close();
```
Comment (martinfurmanski) [inline enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/v2/decoding/DecodingDispatcher.java:63]: Seems like a lifecycle issue to have a decoder in the pipeline that needs to be externally closed. Is it really necessary? Can it be redesigned? Perhaps we can use another Netty handler to clean up resources, i.e. channelInactive()?

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/v2/encoding/RaftMessageContentEncoder.java [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+        this.serializer = serializer;
+    }
+
+    @Override
+    protected void encode( ChannelHandlerContext ctx, RaftMessages.ClusterIdAwareMessage msg, List<Object> out ) throws Exception
+    {
+        out.add( ContentType.Message );
+        out.add( msg );
+        Object[] dispatch = msg.message().dispatch( new Handler() );
+        if ( dispatch == null )
```
Comment (martinfurmanski) [inline enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/v2/encoding/RaftMessageContentEncoder.java]: Can this case really happen?

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/v2/encoding/RaftMessageContentEncoder.java [hunk 13]
```diff
@@ -0,0 +1,173 @@
...
+        {
+            // there was an error serializing
+            throw new IllegalArgumentException( "Error reading raft message content" );
+        }
+        else if ( dispatch.length != 0 )
```
Comment (martinfurmanski) [inline enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/v2/encoding/RaftMessageContentEncoder.java]: List#addAll should already handle the length=0 case cleanly, so no need for the conditional.

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/v2/encoding/RaftMessageContentEncoder.java [hunk 14]
```diff
@@ -0,0 +1,173 @@
...
+    {
+        this.serializer = serializer;
+    }
+
+    @Override
+    protected void encode( ChannelHandlerContext ctx, RaftMessages.ClusterIdAwareMessage msg, List<Object> out ) throws Exception
+    {
+        out.add( ContentType.Message );
+        out.add( msg );
+        Object[] dispatch = msg.message().dispatch( new Handler() );
```
Comment (martinfurmanski) [inline enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/v2/encoding/RaftMessageContentEncoder.java]: It doesn't look like you need to instantiate a new Handler here every time.
Dispatch also seems like an odd variable name.

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/v2/encoding/RaftMessageContentEncoder.java [hunk 15]
```diff
@@ -0,0 +1,173 @@
...
+
+import org.neo4j.causalclustering.core.consensus.RaftMessages;
+import org.neo4j.causalclustering.core.consensus.log.RaftLogEntry;
+import org.neo4j.causalclustering.core.replication.ReplicatedContent;
+import org.neo4j.causalclustering.messaging.marshalling.CoreReplicatedContentSerializer;
+import org.neo4j.causalclustering.messaging.marshalling.v2.ContentType;
+
+import static org.neo4j.causalclustering.messaging.marshalling.v2.encoding.RaftLogEntryTermEncoder.serializable;
+
+public class RaftMessageContentEncoder extends MessageToMessageEncoder<RaftMessages.ClusterIdAwareMessage>
```
Comment (martinfurmanski) [inline enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/v2/encoding/RaftMessageContentEncoder.java]: Need some JavaDoc explaining what the purpose of this encoder is and what it does.

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/v2/encoding/RaftMessageContentEncoder.java [hunk 16]
```diff
@@ -0,0 +1,173 @@
...
+    public RaftMessageContentEncoder( CoreReplicatedContentSerializer serializer )
+    {
+        this.serializer = serializer;
+    }
+
+    @Override
+    protected void encode( ChannelHandlerContext ctx, RaftMessages.ClusterIdAwareMessage msg, List<Object> out ) throws Exception
+    {
+        out.add( ContentType.Message );
+        out.add( msg );
```
Comment (martinfurmanski) [inline enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/v2/encoding/RaftMessageContentEncoder.java]: Seems a bit odd to always pass the msg along, alongside the subsequent transformation. I think this is a bit of spaghetti design. The roles between RaftMessageContentEncoder and RaftMessageEncoder have to be clearer. I suggest to start with renaming RaftMessageContentEncoder to RaftContentEncoder, clarifying its role in the javadoc and make it either handle a class of messages or pass the message along to the next handler, not both.

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/v2/encoding/RaftMessageEncoder.java [hunk 17]
```diff
@@ -0,0 +1,180 @@
...
+        RaftMessages.RaftMessage message = decoratedMessage.message();
+        ClusterId clusterId = decoratedMessage.clusterId();
+        MemberId.Marshal memberMarshal = new MemberId.Marshal();
+
+        NetworkFlushableByteBuf channel = new NetworkFlushableByteBuf( out );
+        ClusterId.Marshal.INSTANCE.marshal( clusterId, channel );
+        channel.putInt( message.type().ordinal() );
+        memberMarshal.marshal( message.from(), channel );
+
+        message.dispatch( new Handler( memberMarshal, channel ) );
```
Comment (martinfurmanski) [inline enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/v2/encoding/RaftMessageEncoder.java:48]: I think you should be able to instantiate this Handler on creation of the RafeMessageEncoder.

**Review by martinfurmanski — CHANGES_REQUESTED** [review]

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/CoreReplicatedContentSerializer.java (hunk 2 above)
Comment (RagnarW) [reply enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/CoreReplicatedContentSerializer.java]: The marshal class was deleted quite late, both were having around for quite some time. I don't mind changing it.

File: enterprise/causal-clustering/src/test/java/org/neo4j/causalclustering/core/consensus/log/RaftContentByteBufferMarshalTest.java (hunk 5 above)
Comment (RagnarW) [reply enterprise/causal-clustering/src/test/java/org/neo4j/causalclustering/core/consensus/log/RaftContentByteBufferMarshalTest.java:55]: Yes. CoreReplicatedContentMarshallingTest

And in that test, all types are tested.

Comment (RagnarW) [reply enterprise/causal-clustering/src/test/java/org/neo4j/causalclustering/core/consensus/log/RaftContentByteBufferMarshalTest.java:55]: Perhaps keep txSerializationShouldNotResultInExcessZeroes and move to ReplicateTransctionFactoryTest

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/ChunkedReplicatedContent.java [hunk 18]
```diff
@@ -0,0 +1,132 @@
...
+                lastByteWasWritten = true;
+            }
+            if ( isEndOfInput() != endOfInput )
+            {
+                buffer.setBoolean( 0, isEndOfInput() );
+            }
+            progress += buffer.readableBytes();
+            return buffer;
+        }
+        catch ( IOException e )
```
Comment (martinfurmanski) [inline enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/ChunkedReplicatedContent.java]: Catch Throwable here? We want to lean up always right?

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/ChunkedReplicatedContent.java [hunk 19]
```diff
@@ -0,0 +1,132 @@
...
+    private final ByteBufAwareMarshal byteBufAwareMarshal;
+    private final int chunkSize;
+    private boolean lastByteWasWritten;
+    private int progress;
+
+    public ChunkedReplicatedContent( byte contentType, ByteBufAwareMarshal byteBufAwareMarshal, int chunkSize )
+    {
+        this.byteBufAwareMarshal = byteBufAwareMarshal;
+        this.chunkSize = chunkSize;
+        if ( chunkSize < 7 )
```
Comment (martinfurmanski) [inline enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/ChunkedReplicatedContent.java]: 7 vs 4 below

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/ChunkedReplicatedContent.java [hunk 20]
```diff
@@ -0,0 +1,132 @@
...
+    public void marshal( WritableChannel channel ) throws IOException
+    {
+        channel.put( contentType );
+        byteBufAwareMarshal.marshal( channel );
+    }
+
+    @Override
+    public boolean isEndOfInput()
+    {
+        return lastByteWasWritten;
```
Comment (martinfurmanski) [inline enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/ChunkedReplicatedContent.java]: Maybe just call this endOfInput?

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/ChunkedReplicatedContent.java [hunk 21]
```diff
@@ -0,0 +1,132 @@
...
+    @Override
+    public ByteBuf readChunk( ChannelHandlerContext ctx ) throws IOException
+    {
+        return readChunk( ctx.alloc() );
+    }
+
+    @Override
+    public ByteBuf readChunk( ByteBufAllocator allocator ) throws IOException
+    {
+        boolean endOfInput = isEndOfInput();
```
Comment (martinfurmanski) [inline enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/ChunkedReplicatedContent.java]: Why not just use the lastBytesWasWritten field?

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/ChunkedReplicatedContent.java [hunk 22]
```diff
@@ -0,0 +1,132 @@
...
+            {
+                // extra metadata on first chunk
+                buffer.writeByte( contentType );
+                buffer.writeInt( byteBufAwareMarshal.length() );
+            }
+            if ( !byteBufAwareMarshal.encode( buffer ) )
+            {
+                lastByteWasWritten = true;
+            }
+            if ( isEndOfInput() != endOfInput )
```
Comment (martinfurmanski) [inline enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/ChunkedReplicatedContent.java]: This is really hard to read because of the overloaded usage with isEndOfInput() vs endOfInput vs lastByteWasWritten. Feels like it could be an else-clause, but I am not even certain what the logic is trying to achieve.

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/CoreReplicatedContentMarshal.java [hunk 23]
```diff
@@ -0,0 +1,156 @@
...
+import org.neo4j.causalclustering.core.state.machines.tx.ReplicatedTransactionSerializer;
+import org.neo4j.causalclustering.core.state.storage.SafeChannelMarshal;
+import org.neo4j.causalclustering.messaging.EndOfStreamException;
+import org.neo4j.storageengine.api.ReadableChannel;
+import org.neo4j.storageengine.api.WritableChannel;
+
+import static java.util.Collections.singleton;
+import static org.neo4j.causalclustering.messaging.marshalling.ByteBufAwareMarshal.simple;
+
+public class CoreReplicatedContentMarshal extends SafeChannelMarshal<ReplicatedContent>
```
Comment (martinfurmanski) [inline enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/CoreReplicatedContentMarshal.java:51]: Are these changes compatible with Raft V1?

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/v2/ContentType.java [hunk 24]
```diff
@@ -0,0 +1,43 @@
...
+package org.neo4j.causalclustering.messaging.marshalling.v2;
+
+public enum ContentType
+{
+    ContentType( (byte) 0 ),
+    ReplicatedContent( (byte) 1 ),
+    RaftLogEntryTerms( (byte) 2 ),
+    Message( (byte) 3 );
+
+    private final byte b;
```
Comment (martinfurmanski) [inline enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/v2/ContentType.java]: Maybe give it a slightly better name: messageCode?

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/v2/decoding/ContentTypeDispatcher.java [hunk 25]
```diff
@@ -0,0 +1,69 @@
...
+    {
+        this.contentTypeProtocol = contentTypeProtocol;
+    }
+
+    @Override
+    protected void decode( ChannelHandlerContext ctx, ByteBuf in, List<Object> out )
+    {
+        if ( contentTypeProtocol.isExpecting( ContentType.ContentType ) )
+        {
+            byte b = in.readByte();
```
Comment (martinfurmanski) [inline enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/v2/decoding/ContentTypeDispatcher.java]: same: messageCode

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/v2/decoding/ContentTypeDispatcher.java [hunk 26]
```diff
@@ -0,0 +1,69 @@
...
+        }
+        else
+        {
+            ctx.fireChannelRead( in.retain() );
+        }
+    }
+
+    private ContentType getContentType( byte b )
+    {
+        for ( ContentType contentType : ContentType.values() )
```
Comment (martinfurmanski) [inline enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/v2/decoding/ContentTypeDispatcher.java]: Seems odd to use a for-loop here. You can access using values()[typeOrdinal]. Perhaps the AIIOB exception is just as good as the IllegalArgumentException.

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/v2/decoding/RaftMessageComposer.java [hunk 27]
```diff
@@ -0,0 +1,112 @@
...
+import java.util.Queue;
+import java.util.concurrent.LinkedBlockingQueue;
+
+import org.neo4j.causalclustering.core.consensus.RaftMessages;
+import org.neo4j.causalclustering.core.replication.ReplicatedContent;
+
+public class RaftMessageComposer extends MessageToMessageDecoder<Object>
+{
+    private final Queue<ReplicatedContent> replicatedContents = new LinkedBlockingQueue<>();
+    private final Queue<Long> raftLogEntryTerms = new LinkedBlockingQueue<>();
```
Comment (martinfurmanski) [inline enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/v2/decoding/RaftMessageComposer.java]: Is there any reason of synchronisation why these need to be LinkedBlockingQueues? Otherwise I suggest some simple Lists. Remember that handlers in a netty pipeline do not run concurrently.

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/v2/decoding/RaftMessageDecoder.java [hunk 28]
```diff
@@ -0,0 +1,230 @@
...
+    {
+        ReadableChannel channel = new NetworkReadableClosableChannelNetty4( buffer );
+        ClusterId clusterId = ClusterId.Marshal.INSTANCE.unmarshal( channel );
+
+        int messageTypeWire = channel.getInt();
+        RaftMessages.Type[] values = RaftMessages.Type.values();
+        RaftMessages.Type messageType = values[messageTypeWire];
+
+        MemberId from = retrieveMember( channel );
+        BiFunction<Queue<Long>,Queue<ReplicatedContent>,RaftMessages.BaseRaftMessage> result;
```
Comment (martinfurmanski) [inline enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/v2/decoding/RaftMessageDecoder.java]: I think it would be good to extract this BiFunction into a proper interface of its own to add some semantic meaning to it. Should probably live somewhere close to RaftMessageCreator.

I think the output type should be RaftMessage and not BaseRaftMessage.

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/v2/decoding/RaftMessageDecoder.java [hunk 29]
```diff
@@ -0,0 +1,230 @@
...
+        else if ( messageType.equals( APPEND_ENTRIES_REQUEST ) )
+        {
+            // how many
+            long term = channel.getLong();
+            long prevLogIndex = channel.getLong();
+            long prevLogTerm = channel.getLong();
+            long leaderCommit = channel.getLong();
+            int raftLogEntries = channel.getInt();
+
+            result = ( rle, rc ) ->
```
Comment (martinfurmanski) [inline enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/v2/decoding/RaftMessageDecoder.java]: Extract this somewhere and give it a name.

**Review by martinfurmanski — COMMENTED** [review]
A few more comments. I also think we need (at least) an IT showing that a cluster using Raft V1 is able to communicate.

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/ChunkedReplicatedContent.java (hunk 21 above)
Comment (RagnarW) [reply enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/ChunkedReplicatedContent.java]: Since the method is part of this class api I think it makes more sense. Internal structure may change but the interface should stay the same.

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/ChunkedReplicatedContent.java (hunk 19 above)
Comment (RagnarW) [reply enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/ChunkedReplicatedContent.java]: Oops

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/CoreReplicatedContentMarshal.java (hunk 23 above)
Comment (RagnarW) [reply enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/CoreReplicatedContentMarshal.java:51]: Yes. We need to verify through an upgrade test or similar. I did test before removing the old class.

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/v2/decoding/ContentTypeDispatcher.java (hunk 26 above)
Comment (RagnarW) [reply enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/v2/decoding/ContentTypeDispatcher.java]: Ordinal does not have to be the same as byte value. We could rely on ordinal if we want.

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/v2/decoding/RaftMessageComposer.java (hunk 27 above)
Comment (RagnarW) [reply enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/v2/decoding/RaftMessageComposer.java]: No reason. The queue interface made sense to me because the decoding relies on FIFO for all the components. I was probably just lazy in picking this queue. Is there a better one in java?

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/ChunkedReplicatedContent.java (hunk 22 above)
Comment (RagnarW) [reply enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/ChunkedReplicatedContent.java]: So logic is. If, after writing to buffer, isEndOfInput has changed then we overwrite that value. Otherwise we leave it. But since all of the logic around this is handled in the same method I can agree that it could be simplified. Or just put the boolean last instead.

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/v2/decoding/RaftMessageDecoder.java (hunk 28 above)
Comment (RagnarW) [reply enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/v2/decoding/RaftMessageDecoder.java]: I think I understand what you mean. I'll show you a suggestion when I have one.

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/v2/decoding/RaftMessageComposer.java (hunk 27 above)
Comment (RagnarW) [reply enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/v2/decoding/RaftMessageComposer.java]: Switched to linked list

File: enterprise/causal-clustering/src/test/java/org/neo4j/causalclustering/catchup/tx/ChunkedReplicatedTransactionInputStreamTest.java (hunk 6 above)
Comment (RagnarW) [reply enterprise/causal-clustering/src/test/java/org/neo4j/causalclustering/catchup/tx/ChunkedReplicatedTransactionInputStreamTest.java]: This was unitentional

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/v2/decoding/DecodingDispatcher.java (hunk 11 above)
Comment (RagnarW) [reply enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/v2/decoding/DecodingDispatcher.java:63]: It can be separated. It was perhaps an short cut.

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/Serializer.java (hunk 10 above)
Comment (RagnarW) [reply enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/Serializer.java]: not if it is to follow the documented intended use of the method. This interface is suppose to be exhausted.

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/Serializer.java (hunk 9 above)
Comment (RagnarW) [reply enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/Serializer.java]: I believe that is explained by the interface. But I guess the description is in the return. I could move it up.

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/Marshal.java (hunk 8 above)
Comment (RagnarW) [reply enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/Marshal.java:29]: It seem you answered your own question. But perhaps this is naming thing. It is not very compatible with the current interface.

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/v2/encoding/RaftMessageContentEncoder.java (hunk 12 above)
Comment (RagnarW) [reply enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/messaging/marshalling/v2/encoding/RaftMessageContentEncoder.java]: Not really
